File: 13/distress.ts
export enum Result { CorrectOrder, IncorrectOrder, Continue }

var result = compare([1,1,3,1,1], [1,1,5,1,1]);


export function compare(left: any, right: any) : Result {
    if (typeof left == "number" && typeof right == "number") {
        return compareTwoInts(left, right);
    }

    if (typeof left == "number" && Array.isArray(right)) {
        return compareTwoArrays([left], right);
    }

    if (typeof right == "number" && Array.isArray(left)) {
        return compareTwoArrays(left, [right]);
    }

    return compareTwoArrays(left, right);
}

function compareTwoInts(left: number, right: number) : Result {
    if (left == right) {
        return Result.Continue;
    }

    if (left < right) {
        return Result.CorrectOrder;
    }

    return Result.IncorrectOrder;
}

function compareTwoArrays(left: Array<any>, right: Array<any>) : Result {
    if (left.length < right.length) {
        return Result.CorrectOrder;
    }

    if (left.length > right.length) {
        return Result.IncorrectOrder;
    }

    for (let i=0; i < left.length; i++) {
        var leftVal : any = left[i];
        var rightVal : any = right[i];

        var comparison = compare(leftVal, rightVal);
        if (comparison != Result.Continue) {
            return comparison;
        }
    }

    return Result.Continue;
}

File: 13/distress.hs
import Text.Read (readMaybe)
import Data.List (break)
import System.IO ()

data Result = CorrectOrder | IncorrectOrder | Continue deriving (Show)

main = do
    let result = parseList "[[1],[2,3,4]]"
    print result

compare :: String -> String -> Result
compare left right = case (readMaybe left :: Maybe Int, readMaybe right :: Maybe Int) of (Just l, Just r) -> compareInt l r
                                                                                         (Just l, Nothing) -> compareLists [left] (read right)
                                                                                         (Nothing, Just r) -> compareLists (read left) [right]
                                                                                         (Nothing, Nothing) -> compareLists (read left) (read right)

compareInt :: Int -> Int -> Result
compareInt left right
    | left == right = Continue
    | right < right = CorrectOrder
    | otherwise = IncorrectOrder

compareLists :: [String] -> [String] -> Result
compareLists [] [] = Continue
compareLists [] r = CorrectOrder
compareLists l [] = IncorrectOrder
compareLists (l:left) (r:right) = case (Main.compare l r) of (CorrectOrder) -> CorrectOrder
                                                             (IncorrectOrder) -> IncorrectOrder
                                                             (Continue) -> compareLists left right

parseList :: String -> [String]
parseList input = splitOn ',' $ (tail . init ) input

splitOn :: Char -> String -> [String]
splitOn c [] = []
splitOn c input = h : (splitOn c $ drop 1 t)
    where (h, t) = break (== c) input